File: backend/internal/scheduler/rr.go
package scheduler

import (
	"sort"

	"github.com/pdro1812/trabalho-os-scheduler/backend/internal/models"
)

func RunRoundRobin(req models.SimulationRequest) models.SimulationResponse {
	processes := make([]models.Process, len(req.Processes))
	copy(processes, req.Processes)

	// Ordena inicialmente por tempo de chegada
	sort.SliceStable(processes, func(i, j int) bool {
		return processes[i].ArrivalTime < processes[j].ArrivalTime
	})

	n := len(processes)
	remainingTime := make([]int, n)
	for i := range processes {
		remainingTime[i] = processes[i].BurstTime
	}

	var executionOrder []models.ExecutionBlock
	var metrics []models.ProcessMetrics

	currentTime := 0
	completed := 0
	queue := []int{}
	arrivalIdx := 0

	// Nova função blindada: Adiciona na fila apenas quem chegou no tempo atual e ainda não entrou
	checkArrivals := func() {
		for arrivalIdx < n && processes[arrivalIdx].ArrivalTime <= currentTime {
			queue = append(queue, arrivalIdx)
			arrivalIdx++
		}
	}

	// Adiciona os processos que chegam no tempo 0
	checkArrivals()
	lastPID := ""

	totalWaitTime := 0
	totalTurnaroundTime := 0

	for completed < n {
		// Se a fila está vazia, a CPU fica ociosa até o próximo processo chegar
		if len(queue) == 0 {
			if arrivalIdx < n {
				currentTime = processes[arrivalIdx].ArrivalTime
				checkArrivals()
			}
			continue
		}

		// Retira o primeiro da fila
		idx := queue[0]
		queue = queue[1:]
		p := processes[idx]

		// Aplica TTC se mudou de processo
		if lastPID != "" && lastPID != p.PID && req.TTC > 0 {
			executionOrder = append(executionOrder, models.ExecutionBlock{
				PID:   "TTC",
				Start: currentTime,
				End:   currentTime + req.TTC,
			})
			currentTime += req.TTC
			checkArrivals() // Verifica se alguém chegou DENTRO do tempo de troca de contexto
		}

		// Define quanto tempo vai rodar (o Quantum ou o que sobrou)
		runTime := req.Quantum
		if remainingTime[idx] < runTime {
			runTime = remainingTime[idx]
		}

		start := currentTime
		currentTime += runTime
		remainingTime[idx] -= runTime

		executionOrder = append(executionOrder, models.ExecutionBlock{
			PID:   p.PID,
			Start: start,
			End:   currentTime,
		})

		// Verifica quem chegou ENQUANTO esse processo rodava na CPU
		checkArrivals()

		// Se o processo ainda não terminou, volta pro final da fila
		if remainingTime[idx] > 0 {
			queue = append(queue, idx)
		} else {
			// Processo terminou
			completed++
			turnaround := currentTime - p.ArrivalTime
			wait := turnaround - p.BurstTime
			totalTurnaroundTime += turnaround
			totalWaitTime += wait

			metrics = append(metrics, models.ProcessMetrics{
				PID:           p.PID,
				EffectiveTime: turnaround,
				WaitingTime:   wait,
			})
		}
		lastPID = p.PID
	}

	return models.SimulationResponse{
		ExecutionOrder:    executionOrder,
		Metrics:           metrics,
		AvgWaitTime:       float64(totalWaitTime) / float64(n),
		AvgTurnaroundTime: float64(totalTurnaroundTime) / float64(n),
	}
}

File: backend/internal/scheduler/rr.js
function runRoundRobin(requisicao) {
    const processos = [];
    for (let i = 0; i < requisicao.Processes.length; i++) {
        processos.push(requisicao.Processes[i]);
    }

    // Ordenamos por tempo de chegada para saber quem entra primeiro na fila
    processos.sort((a, b) => a.ArrivalTime - b.ArrivalTime);

    const n = processos.length;
    const tempoRestante = [];
    for (let i = 0; i < n; i++) {
        tempoRestante.push(processos[i].BurstTime);
    }

    const ordemExecucao = [];
    const metricas = [];
    let tempoAtual = 0;
    let concluídos = 0;
    let somaEspera = 0;
    let somaTurnaround = 0;

    // Fila de prontos (guarda os índices dos processos)
    const fila = [];
    let proximoChegando = 0;

    // Controle de troca de contexto
    let veioDeOciosidade = true;
    let ultimoPID = "";

    // Função simples para colocar na fila quem acabou de chegar
    function checarChegadas() {
        while (proximoChegando < n && processos[proximoChegando].ArrivalTime <= tempoAtual) {
            fila.push(proximoChegando);
            proximoChegando++;
        }
    }

    // Começamos checando quem chega no tempo 0
    checarChegadas();

    while (concluídos < n) {
        
        // Se a fila estiver vazia, a CPU fica ociosa
        if (fila.length === 0) {
            if (proximoChegando < n) {
                tempoAtual = processos[proximoChegando].ArrivalTime;
                veioDeOciosidade = true;
                ultimoPID = "";
                checarChegadas();
            }
            continue;
        }

        // Tiramos o primeiro processo da fila
        const indice = fila.shift();
        const p = processos[indice];

        // --- TROCA DE CONTEXTO ---
        // Se mudou o processo e não veio de ociosidade, cobra TTC
        if (veioDeOciosidade === false && ultimoPID !== "" && ultimoPID !== p.PID && requisicao.TTC > 0) {
            ordemExecucao.push({
                PID: "TTC",
                Start: tempoAtual,
                End: tempoAtual + requisicao.TTC
            });
            tempoAtual = tempoAtual + requisicao.TTC;
            
            // Durante o TTC, outros processos podem ter chegado
            checarChegadas();
        }

        veioDeOciosidade = false;

        // --- EXECUÇÃO (QUANTUM) ---
        // O processo roda o valor do Quantum OU o que resta dele (o que for menor)
        const tempoParaRodar = Math.min(requisicao.Quantum, tempoRestante[indice]);

        const inicio = tempoAtual;
        tempoAtual = tempoAtual + tempoParaRodar;
        tempoRestante[indice] = tempoRestante[indice] - tempoParaRodar;

        ordemExecucao.push({
            PID: p.PID,
            Start: inicio,
            End: tempoAtual
        });
        ultimoPID = p.PID;

        // --- CHECAR NOVAS CHEGADAS ---
        // Importante: quem chega enquanto o processo rodava entra na fila ANTES
        // de quem foi interrompido pelo Quantum (preemptado)
        checarChegadas();

        if (tempoRestante[indice] === 0) {
            // Processo terminou
            concluídos++;
            const turnaround = tempoAtual - p.ArrivalTime;
            const espera = turnaround - p.BurstTime;

            metricas.push({
                PID: p.PID,
                EffectiveTime: turnaround,
                WaitingTime: espera
            });

            somaTurnaround += turnaround;
            somaEspera += espera;
        } else {
            // Processo ainda tem tempo, volta para o fim da fila
            fila.push(indice);
        }
    }

    return {
        ExecutionOrder: ordemExecucao,
        Metrics: metricas,
        AvgWaitTime: somaEspera / n,
        AvgTurnaroundTime: somaTurnaround / n
    };
}

module.exports = { runRoundRobin };